### PR TITLE
Add compass HUD element

### PR DESCRIPTION
**Pull request by EvidenceBKidscode**
This PR adds a "compass" HUD element (new HUD feature).

Basically it is like image HUD element but it rotates or translates according to the current heading direction. This allows to build various compass HUD. The display is entirely updated on client side so it is quite smooth.

## To do

This PR is ready for review.

## How to test

Here is a sample code for testing it, using default mod textures :

```lua
minetest.register_on_joinplayer(function(player)
	player:hud_add({
		hud_elem_type = "compass",
		position = { x = 0, y = 0 },
		offset = { x = 10, y = 10 },
		size = { x = 150, y = 150 },
		alignment = { x = 1, y = 1 },
		text = "gui_furnace_arrow_bg.png",
		dir = 1,
	})

	player:hud_add({
		hud_elem_type = "compass",
		position = { x = 0, y = 0 },
		offset = { x = 10, y = 170 },
		size = { x = 100, y = 50 },
		scale = { x = 1, y = 1 }, -- Required or scale would be 0
		alignment = { x= 1, y = 1 },
		text = "default_ladder_steel.png^[transformR90",
		dir = 2,
	})
end)
```

This adds two compass HUD elements in the upper left corner of the screen (better texture would be much more suitable :smile:) :
screenshot_20200114_164400

### CODE REVIEW

**Comment by LoneWolfHT** [comment]
I support the idea, but am not sure how the second HUD works

**Comment by pyrollo** [comment]
> I don`t understand why the steel ladder element is needed for the compass on the screen?

This was just provided as an example, I could have put anything else.
Imagine putting your own texture instead. Then it'll be up to you to create various compass, either rotating compass or horizontal bar compass.

File: src/hud.h [hunk 1]
```diff
@@ -61,6 +61,7 @@ enum HudElementType {
 	HUD_ELEM_STATBAR   = 2,
 	HUD_ELEM_INVENTORY = 3,
 	HUD_ELEM_WAYPOINT  = 4,
+	HUD_ELEM_COMPASS   = 6,
```
Comment (nerzhul) [inline src/hud.h]: why 6 and not 5 ?

**Comment by Andrey2470T** [comment]
As I consider if you want to know along which horizontal direction (North/South, East/West) you are currently moving, you can just press F5 and you above see info about it. I don`t understand a point of this feature.

**Comment by style-nine** [comment]
Not everyone uses F5 for gameplay. Some people reserve that strictly for debugging.

File: src/hud.h (hunk 1 above)
Comment (SmallJoker) [reply src/hud.h]: 5 will be `minimap`.
https://github.com/minetest/minetest/pull/9079/files#diff-036e439b344795cd3c89ffb3153b05d7R64

Comment (pyrollo) [reply src/hud.h]: Thanks SmallJocker, that's it.
I was hesitating between having same number in the two PRs then reajust, or try to hide this detail from perspicacious eyes of core devs :)

**Comment by Andrey2470T** [comment]
I think it would be nice to embed default compass and reserve for it a special key (e.g F11 or any else) to be able to activate/deactivate it. Also, shift it to right or left downer corner as it would overlap the chat text. But probably this is already other topic.

**Comment by pyrollo** [comment]
In my opinion, adding default stuff to engine is a mistake. Specific things like compass (an minimap) should be managed by game and mods. This gives much more freedom to modders to do what they want.
The given example is just an example, you can easily change the position to whatever you want, there is no default position in this PR.

**Comment by nerzhul** [comment]
i agree with @pyrollo .
Rebase is needed here. Will do some review stuff after

**Comment by pyrollo** [comment]
Rebased

File: doc/lua_api.txt [hunk 2]
```diff
@@ -1404,7 +1404,23 @@ Same as `image`, but does not accept a `position`; the position is instead deter
...
 
+Displays an image oriented or translated according to current heading direction.
+
+* `size`: The size of this element.  Negative values represent that percentage
+  of the screen it should take; e.g. `x=-100` means 100% (width).
+* `scale`: Scale of the translated image (useless if rotation selected).
+* `text`: The name of the texture that is displayed.
+* `alignment`: The alignment of the image.
+* `offset`: offset in pixels from position.
+* `dir`: How image is rotated/translated :
```
Comment (SmallJoker) [inline doc/lua_api.txt:1417]:
```suggestion
* `dir`: How the image is rotated/translated:
```

File: doc/lua_api.txt [hunk 3]
```diff
@@ -1404,7 +1404,23 @@ Same as `image`, but does not accept a `position`; the position is instead deter
 * `world_pos`: World position of the waypoint.
 * `offset`: offset in pixels from position.
 
+### `compass`
 
+Displays an image oriented or translated according to current heading direction.
+
+* `size`: The size of this element.  Negative values represent that percentage
+  of the screen it should take; e.g. `x=-100` means 100% (width).
+* `scale`: Scale of the translated image (useless if rotation selected).
```
Comment (SmallJoker) [inline doc/lua_api.txt:1413]: When is this parameter needed, when is it superfluous? The phrasing is unclear to me.

File: src/client/hud.cpp [hunk 4]
```diff
@@ -393,13 +415,131 @@ void Hud::drawLuaElements(const v3s16 &camera_offset)
...
+				// Limit angle and ajust with given offset
+				angle = (angle + (int)e->number) % 360;
+
+				core::rect<s32> dstrect(0, 0, dstsize.X, dstsize.Y);
+				dstrect += pos + v2s32(
+								(e->align.X - 1.0) * dstsize.X / 2,
+								(e->align.Y - 1.0) * dstsize.Y / 2) +
+						v2s32(e->offset.X * m_hud_scaling, e->offset.Y * m_hud_scaling);
+
+				switch(e->dir) {
```
Comment (SmallJoker) [inline src/client/hud.cpp:448]:
```suggestion
				switch (e->dir) {
```

File: doc/lua_api.txt (hunk 3 above)
Comment (pyrollo) [reply doc/lua_api.txt]: Tried to make it clearer :)

**Comment by paramat** [comment]
As pyrollo is now a core dev, this can be merged.

Although, i am unsure this is worth adding.
A compass does not seem much needed and we already have degrees and NESW readouts in F5 for when it is needed.
A compass seems a thing you only occasionally need, so always being on-screen seems obtrusive, in this way F5 seems better.
Although i agree that a compass feature being controlled from server, games and mods is better than client-controlled.

PR #9315 hides yaw according to a privilege, perhaps this is all we need as it makes it server controlled?

**Comment by rubenwardy** [comment]
> we already have degrees and NESW readouts in F5 for when it is needed

All games could just be text adventures, but that would be no fun

**Comment by paramat** [comment]
Well, that is just a snarky comment that misrepresents my comment to ridicule it.

**Comment by pyrollo** [comment]
I think Rubenwardy is right, visual effects are important in video games, they improve immersion. This PR is less about showing the heading direction than adding visual effects.

One criticism I could make to my own PR is that it is a bit limited. Such elements could also be able to point to a map point or something else. But HUD specification was too limited to do that. Maybe with #10245 it could be possible to have a very versatile "moving or rotating" image HUD element.

Example use cases:
- Compass (rotating or vertical + horizontal);
- Kind of radar indicating direction of a place;
- Another kind of radar indicating directions of mobs;

So should this PR be improved before merged ? (I guess yes)

**Pull request merged**